crates/farming/src/lib.rs

```diff
@@ -0,0 +1,357 @@
+                    pool_currency_id,
+                    reward_currency_id,
+                    period_count: total_period_count,
+                    per_period: total_per_period,
+                });
+                Ok(().into())
+            })
+        }
+
+        /// Explicitly remove a reward schedule and transfer any remaining
+        /// balance to the treasury
+        #[pallet::weight(T::WeightInfo::remove_reward_schedule())]
+        #[transactional]
+        pub fn remove_reward_schedule(
+            origin: OriginFor<T>,
+            pool_currency_id: CurrencyIdOf<T>,
+            reward_currency_id: CurrencyIdOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            ensure_root(origin)?;
+
+            // transfer unspent rewards to treasury
+            let treasury_account_id = Self::treasury_account_id();
+            let pool_account_id = Self::pool_account_id(&pool_currency_id);
+            T::MultiCurrency::transfer(
+                reward_currency_id,
+                &pool_account_id,
+                &treasury_account_id,
+                T::MultiCurrency::free_balance(reward_currency_id, &pool_account_id),
+            )?;
+
+            RewardSchedules::<T>::remove(pool_currency_id, reward_currency_id);
+            Self::deposit_event(Event::RewardScheduleUpdated {
+                pool_currency_id,
+                reward_currency_id,
+                period_count: Zero::zero(),
+                per_period: Zero::zero(),
+            });
+
+            Ok(().into())
+        }
+
+        /// Stake the pool tokens in the reward pool
+        #[pallet::weight(T::WeightInfo::deposit())]
+        #[transactional]
+        pub fn deposit(
+            origin: OriginFor<T>,
+            pool_currency_id: CurrencyIdOf<T>,
+            amount: BalanceOf<T>,
+        ) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+
+            // prevent depositing without reward schedule
+            ensure!(
+                !RewardSchedules::<T>::iter_prefix_values(pool_currency_id)
+                    .count()
+                    .is_zero(),
+                Error::<T>::ScheduleNotFound
+            );
+
+            // reserve lp tokens to prevent spending
+            T::MultiCurrency::reserve(pool_currency_id.clone(), &who, amount)?;
+
+            // deposit lp tokens as stake
+            T::RewardPools::deposit_stake(&pool_currency_id, &who, amount)
+        }
+
+        /// Unstake the pool tokens from the reward pool
+        #[pallet::weight(T::WeightInfo::withdraw())]
+        #[transactional]
+        pub fn withdraw(
+            origin: OriginFor<T>,
+            pool_currency_id: CurrencyIdOf<T>,
+            amount: BalanceOf<T>,
+        ) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+
+            // unreserve lp tokens to allow spending
+            let _remaining = T::MultiCurrency::unreserve(pool_currency_id.clone(), &who, amount);
```

Comment:
Should we not catch this error? I mean, it if withdraw_stake succeeds, then unreserving _should_ succeeds as well, but I see no good reason to explicitly ignore any errors

Reply:
The only time `_remaining` would be non-zero is if `amount > reserved` but since `withdraw_stake` will anyway limit the total amount that can be withdrawn I think we can ignore this check

Reply:
Better safe than sorry?

Reply:
Huh I assumed `unreserve` would return a `Result`, but actually that's not the case. With your updated version you're still ignoring the returned value, technically we should check that the returned value is zero